nixos/modules/virtualisation/amazon-init.nix

```diff
@@ -43,7 +41,18 @@ let
     nixos-rebuild switch
   '';
 in {
-  boot.postBootCommands = ''
-    ${bootScript} &
-  '';
+  systemd.services.amazon-init = {
+    inherit script;
+    description = "Reconfigure the system from EC2 userdata on startup";
+
+    wantedBy = [ "sshd.service" ];
+    after = [ "network-online.target" ];
+    requires = [ "network-online.target" ];
+ 
+    restartIfChanged = false;
+    unitConfig.X-StopOnRemoval = false;
+
+    serviceConfig.Type = "oneshot";
```

Comment:
Just a follow-up to this: semantically, I think it makes sense to think of a unit of this type being in the "started" state to signifiy that the post-condition it establishes holds ("foo is initialized"); if the post-condition continues to hold after `MAINPID` exits, it should `RemainAfterExit`. On a more practical note, I suppose it's more about avoiding unnecessary work (or, worst-case, clobber state for already running services, in case the init is destructive).